src/iso15118/d20/state/power_delivery.cpp

```diff
@@ -38,15 +44,48 @@ void PowerDelivery::enter() {
 
 Result PowerDelivery::feed(Event ev) {
 
+    const auto selected_energy_service = m_ctx.session.get_selected_services().selected_energy_service;
+    const bool ac_charger =
+        selected_energy_service == dt::ServiceCategory::AC or selected_energy_service == dt::ServiceCategory::AC_BPT;
+    const bool dc_charger =
+        selected_energy_service == dt::ServiceCategory::DC or selected_energy_service == dt::ServiceCategory::DC_BPT;
+
     if (ev == Event::CONTROL_MESSAGE) {
-        const auto control_data = m_ctx.get_control_event<PresentVoltageCurrent>();
-        if (not control_data) {
-            // Ignore control message
-            return {};
+
+        if (const auto* control_data = m_ctx.get_control_event<PresentVoltageCurrent>()) {
+            present_voltage = control_data->voltage;
+        } else if (const auto* control_data = m_ctx.get_control_event<ClosedContactor>()) {
+            ac_connector_closed = control_data;
+            m_ctx.stop_timeout(d20::TimeoutType::CONTACTOR);
+
+            if (not previous_req.has_value()) {
+                logf_warning("There is no power_delivery_req messages saved!");
+                return {};
+            }
+
+            const auto& res = handle_request(previous_req.value(), m_ctx.session, false);
+            m_ctx.respond(res);
```

Comment:
Would it make sense here to check if the new `ac_connector_closed` is `true` before canceling the timeout and replying?